File: Examples/test-suite/perl5/cpp11_std_unique_ptr_runme.pl
use strict;
use warnings;
use Test::More tests => 100;
BEGIN { use_ok('cpp11_std_unique_ptr') }
require_ok('cpp11_std_unique_ptr');

# adapted from ../java/cpp11_std_unique_ptr_runme.java

sub checkCount {
  my($expected_count) = @_;
  my $actual_count = cpp11_std_unique_ptr::Klass::getTotal_count();
  is($actual_count, $expected_count, "Counts incorrect, expected: $expected_count actual: $actual_count");
}

# Test raw pointer handling involving virtual inheritance
{
  my $kini = new cpp11_std_unique_ptr::KlassInheritance("KlassInheritanceInput");
  checkCount(1);
  my $s = cpp11_std_unique_ptr::useKlassRawPtr($kini);
  is($s, "KlassInheritanceInput", "Incorrect string: $s");
  undef $kini;
  checkCount(0);
}


# #### INPUT BY VALUE ####
# unique_ptr as input
{
  my $kin = new cpp11_std_unique_ptr::Klass("KlassInput");
  checkCount(1);
  my $s = cpp11_std_unique_ptr::takeKlassUniquePtr($kin);
  checkCount(0);
  is($s, "KlassInput", "Incorrect string: $s");
  is(cpp11_std_unique_ptr::is_nullptr($kin), 1, "is_nullptr check");
  undef $kin; # Should not fail, even though already deleted
  checkCount(0);
}

{
  my $kin = new cpp11_std_unique_ptr::Klass("KlassInput");
  checkCount(1);
  my $s = cpp11_std_unique_ptr::takeKlassUniquePtr($kin);
  checkCount(0);
  is($s, "KlassInput", "Incorrect string: $s");
  is(cpp11_std_unique_ptr::is_nullptr($kin), 1, "is_nullptr check");
  eval {
    cpp11_std_unique_ptr::takeKlassUniquePtr($kin);
  };
  like($@, qr/\bcannot release ownership as memory is not owned\b/, "double usage of takeKlassUniquePtr should be an error");
  undef $kin; # Should not fail, even though already deleted
  checkCount(0);
}

{
  my $kin = new cpp11_std_unique_ptr::Klass("KlassInput");
  my $notowned = cpp11_std_unique_ptr::get_not_owned_ptr($kin);
  eval {
    cpp11_std_unique_ptr::takeKlassUniquePtr($notowned);
  };
  like($@, qr/\bcannot release ownership as memory is not owned\b/, "double usage of takeKlassUniquePtr should be an error");
  checkCount(1);
  undef $kin;
  checkCount(0);
}

{
  my $kini = new cpp11_std_unique_ptr::KlassInheritance("KlassInheritanceInput");
  checkCount(1);
  my $s = cpp11_std_unique_ptr::takeKlassUniquePtr($kini);
  checkCount(0);
  is($s, "KlassInheritanceInput", "Incorrect string: $s");
  is(cpp11_std_unique_ptr::is_nullptr($kini), 1, "is_nullptr failed");
  undef $kini; # Should not fail, even though already deleted
  checkCount(0);
}

cpp11_std_unique_ptr::takeKlassUniquePtr(undef);
cpp11_std_unique_ptr::takeKlassUniquePtr(cpp11_std_unique_ptr::make_null());
checkCount(0);

# overloaded parameters
is(cpp11_std_unique_ptr::overloadTest(), 0, "overloadTest failed");
is(cpp11_std_unique_ptr::overloadTest(undef), 1, "overloadTest failed");
is(cpp11_std_unique_ptr::overloadTest(new cpp11_std_unique_ptr::Klass("over")), 1, "overloadTest failed");
checkCount(0);


# #### INPUT BY RVALUE REF ####
# unique_ptr as input
{
  my $kin = new cpp11_std_unique_ptr::Klass("KlassInput");
  checkCount(1);
  my $s = cpp11_std_unique_ptr::moveKlassUniquePtr($kin);
  checkCount(0);
  is($s, "KlassInput", "Incorrect string: $s");
  is(cpp11_std_unique_ptr::is_nullptr($kin), 1, "is_nullptr check");
  undef $kin; # Should not fail, even though already deleted
  checkCount(0);
}

{
  my $kin = new cpp11_std_unique_ptr::Klass("KlassInput");
  checkCount(1);
  my $s = cpp11_std_unique_ptr::moveKlassUniquePtr($kin);
  checkCount(0);
  is($s, "KlassInput", "Incorrect string: $s");
  is(cpp11_std_unique_ptr::is_nullptr($kin), 1, "is_nullptr check");
  eval {
    cpp11_std_unique_ptr::moveKlassUniquePtr($kin);
  };
  like($@, qr/\bcannot release ownership as memory is not owned\b/, "double usage of moveKlassUniquePtr should be an error");
  undef $kin; # Should not fail, even though already deleted
  checkCount(0);
}

{
  my $kin = new cpp11_std_unique_ptr::Klass("KlassInput");
  my $notowned = cpp11_std_unique_ptr::get_not_owned_ptr($kin);
  eval {
    cpp11_std_unique_ptr::moveKlassUniquePtr($notowned);
  };
  like($@, qr/\bcannot release ownership as memory is not owned\b/, "double usage of moveKlassUniquePtr should be an error");
  checkCount(1);
  undef $kin;
  checkCount(0);
}

{
  my $kini = new cpp11_std_unique_ptr::KlassInheritance("KlassInheritanceInput");
  checkCount(1);
  my $s = cpp11_std_unique_ptr::moveKlassUniquePtr($kini);
  checkCount(0);
  is($s, "KlassInheritanceInput", "Incorrect string: $s");
  is(cpp11_std_unique_ptr::is_nullptr($kini), 1, "is_nullptr failed");
  undef $kini; # Should not fail, even though already deleted
  checkCount(0);
}

cpp11_std_unique_ptr::moveKlassUniquePtr(undef);
cpp11_std_unique_ptr::moveKlassUniquePtr(cpp11_std_unique_ptr::make_null());
checkCount(0);

# overloaded parameters
is(cpp11_std_unique_ptr::moveOverloadTest(), 0, "moveOverloadTest failed");
is(cpp11_std_unique_ptr::moveOverloadTest(undef), 1, "moveOverloadTest failed");
is(cpp11_std_unique_ptr::moveOverloadTest(new cpp11_std_unique_ptr::Klass("over")), 1, "moveOverloadTest failed");
checkCount(0);


# #### INPUT BY NON-CONST LVALUE REF ####
# unique_ptr as input
{
  my $kin = new cpp11_std_unique_ptr::Klass("KlassInput");
  checkCount(1);
  my $s = cpp11_std_unique_ptr::moveRefKlassUniquePtr($kin);
  checkCount(0);
  is($s, "KlassInput", "Incorrect string: $s");
  is(cpp11_std_unique_ptr::is_nullptr($kin), 1, "is_nullptr check");
  undef $kin; # Should not fail, even though already deleted
  checkCount(0);
}

{
  my $kin = new cpp11_std_unique_ptr::Klass("KlassInput");
  checkCount(1);
  my $s = cpp11_std_unique_ptr::moveRefKlassUniquePtr($kin);
  checkCount(0);
  is($s, "KlassInput", "Incorrect string: $s");
  is(cpp11_std_unique_ptr::is_nullptr($kin), 1, "is_nullptr check");
  eval {
    cpp11_std_unique_ptr::moveRefKlassUniquePtr($kin);
  };
  like($@, qr/\bcannot release ownership as memory is not owned\b/, "double usage of moveRefKlassUniquePtr should be an error");
  undef $kin; # Should not fail, even though already deleted
  checkCount(0);
}

{
  my $kin = new cpp11_std_unique_ptr::Klass("KlassInput");
  my $notowned = cpp11_std_unique_ptr::get_not_owned_ptr($kin);
  eval {
    cpp11_std_unique_ptr::moveRefKlassUniquePtr($notowned);
  };
  like($@, qr/\bcannot release ownership as memory is not owned\b/, "double usage of moveRefKlassUniquePtr should be an error");
  checkCount(1);
  undef $kin;
  checkCount(0);
}

{
  my $kini = new cpp11_std_unique_ptr::KlassInheritance("KlassInheritanceInput");
  checkCount(1);
  my $s = cpp11_std_unique_ptr::moveRefKlassUniquePtr($kini);
  checkCount(0);
  is($s, "KlassInheritanceInput", "Incorrect string: $s");
  is(cpp11_std_unique_ptr::is_nullptr($kini), 1, "is_nullptr failed");
  undef $kini; # Should not fail, even though already deleted
  checkCount(0);
}

cpp11_std_unique_ptr::moveRefKlassUniquePtr(undef);
cpp11_std_unique_ptr::moveRefKlassUniquePtr(cpp11_std_unique_ptr::make_null());
checkCount(0);

# overloaded parameters
is(cpp11_std_unique_ptr::moveRefOverloadTest(), 0, "moveRefOverloadTest failed");
is(cpp11_std_unique_ptr::moveRefOverloadTest(undef), 1, "moveRefOverloadTest failed");
is(cpp11_std_unique_ptr::moveRefOverloadTest(new cpp11_std_unique_ptr::Klass("over")), 1, "moveRefOverloadTest failed");
checkCount(0);


# #### INPUT BY CONST LVALUE REF ####
# unique_ptr as input
{
  my $kin = new cpp11_std_unique_ptr::Klass("KlassInput");
  checkCount(1);
  my $s = cpp11_std_unique_ptr::useRefKlassUniquePtr($kin);
  checkCount(1);
  is($s, "KlassInput", "Incorrect string: $s");
  undef $kin;
  checkCount(0);
}

{
  my $kini = new cpp11_std_unique_ptr::KlassInheritance("KlassInheritanceInput");
  checkCount(1);
  my $s = cpp11_std_unique_ptr::useRefKlassUniquePtr($kini);
  checkCount(1);
  is($s, "KlassInheritanceInput", "Incorrect string: $s");
  undef $kini;
  checkCount(0);
}

cpp11_std_unique_ptr::useRefKlassUniquePtr(undef);
cpp11_std_unique_ptr::useRefKlassUniquePtr(cpp11_std_unique_ptr::make_null());
checkCount(0);

# overloaded parameters
is(cpp11_std_unique_ptr::useRefOverloadTest(), 0, "useRefOverloadTest failed");
is(cpp11_std_unique_ptr::useRefOverloadTest(undef), 1, "useRefOverloadTest failed");
my $kin = new cpp11_std_unique_ptr::Klass("over");
is(cpp11_std_unique_ptr::useRefOverloadTest($kin), 1, "useRefOverloadTest failed");
checkCount(1);
undef $kin;
checkCount(0);


# unique_ptr as output
my $k1 = cpp11_std_unique_ptr::makeKlassUniquePtr("first");
my $k2 = cpp11_std_unique_ptr::makeKlassUniquePtr("second");
checkCount(2);

undef $k1;
checkCount(1);

is($k2->getLabel, "second", "proper label");

undef $k2;
checkCount(0);

is(cpp11_std_unique_ptr::makeNullUniquePtr(), undef);

# unique_ptr as output (rvalue ref)
my $k1 = cpp11_std_unique_ptr::makeRVRKlassUniquePtr("first");
is($k1->getLabel, "first", "proper label");
is(cpp11_std_unique_ptr::makeRVRKlassUniquePtr(undef), undef);

# unique_ptr as output (lvalue ref)
my $k1 = cpp11_std_unique_ptr::makeRefKlassUniquePtr("lvalueref");
is($k1->getLabel, "lvalueref", "proper label");
is(cpp11_std_unique_ptr::makeRefKlassUniquePtr(undef), undef);

File: Examples/test-suite/octave/cpp11_std_unique_ptr_runme.m
# do not dump Octave core
if exist("crash_dumps_octave_core", "builtin")
  crash_dumps_octave_core(0);
endif

cpp11_std_unique_ptr

function checkCount(expected_count)
  actual_count = Klass_getTotal_count();
  if (actual_count != expected_count)
    error("Counts incorrect, expected:%d actual:%d", expected_count, actual_count);
  endif
end

# Test raw pointer handling involving virtual inheritance
kini = KlassInheritance("KlassInheritanceInput");
checkCount(1);
s = useKlassRawPtr(kini);
if (!strcmp(s, "KlassInheritanceInput"))
  error("Incorrect string: %s", s);
endif
clear kini;
checkCount(0);


# #### INPUT BY VALUE ####
# unique_ptr as input
kin = Klass("KlassInput");
checkCount(1);
s = takeKlassUniquePtr(kin);
checkCount(0);
if (!strcmp(s, "KlassInput"))
  error("Incorrect string: %s", s);
endif
if (!is_nullptr(kin))
  error("is_nullptr failed");
endif
clear kin; # Should not fail, even though already deleted
checkCount(0);

kin = Klass("KlassInput");
checkCount(1);
s = takeKlassUniquePtr(kin);
checkCount(0);
if (!strcmp(s, "KlassInput"))
  error("Incorrect string: %s", s);
endif
if (!is_nullptr(kin))
  error("is_nullptr failed");
endif
exception_thrown = false;
try
  takeKlassUniquePtr(kin);
catch e
  if (isempty(strfind(e.message, "cannot release ownership as memory is not owned")))
    error("incorrect exception message %s", e.message);
  endif
  exception_thrown = true;
end_try_catch
if (!exception_thrown)
    error("double usage of takeKlassUniquePtr should have been an error");
endif
clear kin; # Should not fail, even though already deleted
checkCount(0);

kin = Klass("KlassInput");
exception_thrown = false;
notowned = get_not_owned_ptr(kin);
try
  takeKlassUniquePtr(notowned);
catch e
  if (isempty(strfind(e.message, "cannot release ownership as memory is not owned")))
    error("incorrect exception message %s", e.message);
  endif
  exception_thrown = true;
end_try_catch
if (!exception_thrown)
  error("Should have thrown 'Cannot release ownership as memory is not owned' error");
endif
checkCount(1);
clear kin;
checkCount(0);

kini = KlassInheritance("KlassInheritanceInput");
checkCount(1);
s = takeKlassUniquePtr(kini);
checkCount(0);
if (!strcmp(s, "KlassInheritanceInput"))
  error("Incorrect string: %s", s);
endif
if (!is_nullptr(kini))
  error("is_nullptr failed");
endif
clear kini; # Should not fail, even though already deleted
checkCount(0);

null = []; # NULL pointer
null_ptr = make_null();
takeKlassUniquePtr([]);
takeKlassUniquePtr(null);
takeKlassUniquePtr(null_ptr);
checkCount(0);

# overloaded parameters
if (overloadTest() != 0)
  error("overloadTest failed");
endif
if (overloadTest(null) != 1)
  error("overloadTest failed");
endif
if (overloadTest(Klass("over")) != 1)
  error("overloadTest failed");
endif
checkCount(0);


# #### INPUT BY RVALUE REF ####
# unique_ptr as input
kin = Klass("KlassInput");
checkCount(1);
s = moveKlassUniquePtr(kin);
checkCount(0);
if (!strcmp(s, "KlassInput"))
  error("Incorrect string: %s", s);
endif
if (!is_nullptr(kin))
  error("is_nullptr failed");
endif
clear kin; # Should not fail, even though already deleted
checkCount(0);

kin = Klass("KlassInput");
checkCount(1);
s = moveKlassUniquePtr(kin);
checkCount(0);
if (!strcmp(s, "KlassInput"))
  error("Incorrect string: %s", s);
endif
if (!is_nullptr(kin))
  error("is_nullptr failed");
endif
exception_thrown = false;
try
  moveKlassUniquePtr(kin);
catch e
  if (isempty(strfind(e.message, "cannot release ownership as memory is not owned")))
    error("incorrect exception message %s", e.message);
  endif
  exception_thrown = true;
end_try_catch
if (!exception_thrown)
    error("double usage of moveKlassUniquePtr should have been an error");
endif
clear kin; # Should not fail, even though already deleted
checkCount(0);

kin = Klass("KlassInput");
exception_thrown = false;
notowned = get_not_owned_ptr(kin);
try
  moveKlassUniquePtr(notowned);
catch e
  if (isempty(strfind(e.message, "cannot release ownership as memory is not owned")))
    error("incorrect exception message %s", e.message);
  endif
  exception_thrown = true;
end_try_catch
if (!exception_thrown)
  error("Should have thrown 'Cannot release ownership as memory is not owned' error");
endif
checkCount(1);
clear kin;
checkCount(0);

kini = KlassInheritance("KlassInheritanceInput");
checkCount(1);
s = moveKlassUniquePtr(kini);
checkCount(0);
if (!strcmp(s, "KlassInheritanceInput"))
  error("Incorrect string: %s", s);
endif
if (!is_nullptr(kini))
  error("is_nullptr failed");
endif
clear kini; # Should not fail, even though already deleted
checkCount(0);

null = []; # NULL pointer
null_ptr = make_null();
moveKlassUniquePtr([]);
moveKlassUniquePtr(null);
moveKlassUniquePtr(null_ptr);
checkCount(0);

# overloaded parameters
if (moveOverloadTest() != 0)
  error("moveOverloadTest failed");
endif
if (moveOverloadTest(null) != 1)
  error("moveOverloadTest failed");
endif
if (moveOverloadTest(Klass("over")) != 1)
  error("moveOverloadTest failed");
endif
checkCount(0);


# #### INPUT BY NON-CONST LVALUE REF ####
# unique_ptr as input
kin = Klass("KlassInput");
checkCount(1);
s = moveRefKlassUniquePtr(kin);
checkCount(0);
if (!strcmp(s, "KlassInput"))
  error("Incorrect string: %s", s);
endif
if (!is_nullptr(kin))
  error("is_nullptr failed");
endif
clear kin; # Should not fail, even though already deleted
checkCount(0);

kin = Klass("KlassInput");
checkCount(1);
s = moveRefKlassUniquePtr(kin);
checkCount(0);
if (!strcmp(s, "KlassInput"))
  error("Incorrect string: %s", s);
endif
if (!is_nullptr(kin))
  error("is_nullptr failed");
endif
exception_thrown = false;
try
  moveRefKlassUniquePtr(kin);
catch e
  if (isempty(strfind(e.message, "cannot release ownership as memory is not owned")))
    error("incorrect exception message %s", e.message);
  endif
  exception_thrown = true;
end_try_catch
if (!exception_thrown)
    error("double usage of moveRefKlassUniquePtr should have been an error");
endif
clear kin; # Should not fail, even though already deleted
checkCount(0);

kin = Klass("KlassInput");
exception_thrown = false;
notowned = get_not_owned_ptr(kin);
try
  moveRefKlassUniquePtr(notowned);
catch e
  if (isempty(strfind(e.message, "cannot release ownership as memory is not owned")))
    error("incorrect exception message %s", e.message);
  endif
  exception_thrown = true;
end_try_catch
if (!exception_thrown)
  error("Should have thrown 'Cannot release ownership as memory is not owned' error");
endif
checkCount(1);
clear kin;
checkCount(0);

kini = KlassInheritance("KlassInheritanceInput");
checkCount(1);
s = moveRefKlassUniquePtr(kini);
checkCount(0);
if (!strcmp(s, "KlassInheritanceInput"))
  error("Incorrect string: %s", s);
endif
if (!is_nullptr(kini))
  error("is_nullptr failed");
endif
clear kini; # Should not fail, even though already deleted
checkCount(0);

null = []; # NULL pointer
null_ptr = make_null();
moveRefKlassUniquePtr([]);
moveRefKlassUniquePtr(null);
moveRefKlassUniquePtr(null_ptr);
checkCount(0);

# overloaded parameters
if (moveRefOverloadTest() != 0)
  error("moveRefOverloadTest failed");
endif
if (moveRefOverloadTest(null) != 1)
  error("moveRefOverloadTest failed");
endif
if (moveRefOverloadTest(Klass("over")) != 1)
  error("moveRefOverloadTest failed");
endif
checkCount(0);


# #### INPUT BY CONST LVALUE REF ####
# unique_ptr as input
kin = Klass("KlassInput");
checkCount(1);
s = useRefKlassUniquePtr(kin);
checkCount(1);
if (!strcmp(s, "KlassInput"))
  error("Incorrect string: %s", s);
endif
clear kin;
checkCount(0);

kini = KlassInheritance("KlassInheritanceInput");
checkCount(1);
s = useRefKlassUniquePtr(kini);
checkCount(1);
if (!strcmp(s, "KlassInheritanceInput"))
  error("Incorrect string: %s", s);
endif
clear kini;
checkCount(0);

null = []; # NULL pointer
null_ptr = make_null();
useRefKlassUniquePtr([]);
useRefKlassUniquePtr(null);
useRefKlassUniquePtr(null_ptr);
checkCount(0);

# overloaded parameters
if (useRefOverloadTest() != 0)
  error("useRefOverloadTest failed");
endif
if (useRefOverloadTest(null) != 1)
  error("useRefOverloadTest failed");
endif
kin = Klass("over");
if (useRefOverloadTest(kin) != 1)
  error("useRefOverloadTest failed");
endif
checkCount(1);
clear kin
checkCount(0);


# unique_ptr as output
k1 = makeKlassUniquePtr("first");
if (!strcmp(k1.getLabel(), "first"))
  error("wrong object label");
endif

k2 = makeKlassUniquePtr("second");
checkCount(2);

clear k1;
checkCount(1);

if (!strcmp(k2.getLabel(), "second"))
  error("wrong object label");
endif

clear k2;
checkCount(0);

null_smart_prt = makeNullUniquePtr();
assert(ismatrix(null_smart_prt))
assert(size(null_smart_prt) == size([]))
assert(isequal([], null_smart_prt))

# unique_ptr as output (rvalue ref)
k1 = makeRVRKlassUniquePtr("first");
if (!strcmp(k1.getLabel(), "first"))
  error("wrong object label");
endif
k1 = makeRVRKlassUniquePtr([]);
assert(isequal([], k1))

# unique_ptr as output (lvalue ref)
k1 = makeRefKlassUniquePtr("lvalueref");
if (!strcmp(k1.getLabel(), "lvalueref"))
  error("wrong object label");
endif
k1 = makeRefKlassUniquePtr([]);
assert(isequal([], k1))
